### PR TITLE
Mod pow spike

### PR DESCRIPTION
**Pull request by metanest**
A new method, Integer#mod_pow.

x.mod_pow(a, m) -> (x**a) % m

### CODE REVIEW

File: numeric.c [hunk 1]
```diff
@@ -3996,6 +3996,157 @@ rb_int_pow(VALUE x, VALUE y)
...
+            long const mm = FIX2LONG(m);
+            if (mm <= half_val) {
+                return int_pow_tmp1(rb_int_modulo(a, m), b, mm, nega_flg);
+            } else {
+                return int_pow_tmp2(rb_int_modulo(a, m), b, mm, nega_flg);
+            }
+        } else if (RB_TYPE_P(m, T_BIGNUM)) {
+            return int_pow_tmp3(rb_int_modulo(a, m), b, m, nega_flg);
+        } else {
+            assert(0);
```
Comment (mrkn) [inline numeric.c]: Use `UNREACHABLE` macro.

**Comment by mrkn** [comment]
@metanest Could you write tests for the new feature?

**Comment by Ptico** [comment]
mod pow often used for crypto, so may be it's better to use mpz_powm_sec?

**Comment by mrkn** [comment]
MEMO: The corresponding issue is https://bugs.ruby-lang.org/issues/12508

**Pull request closed without merging**